### PR TITLE
Fix S4457 FP: Argument check after async code

### LINKED ISSUE #6449: S4457 False positive when argument check is after async code
### Description

In Sonar Analyzer version 8.49, the following code, the method HasS4457Async erroneously raises S4457, whereas this does not happen for the method NoS4457Async

The method cannot be split to do the validation before the async code as the validation relies on the result of the async code.

In Sonar Analyzer version 8.48, it doesn't cause S4457.

### Repro steps

````c#
    public class Class1
    {
        public Task<int[]> CheckAsync() => Task.FromResult(new int[] { 1 });
        public Task<int> Check2Async() => Task.FromResult(1);

        public async Task HasS4457Async(int request)
        {
            var identifierType = (await CheckAsync()).FirstOrDefault(x => x == request);
            if (identifierType == 0) throw new ArgumentException("message");
        }

        public async Task NoS4457Async(int request)
        {
            var identifierType = await Check2Async();
            if (identifierType != request) throw new ArgumentException("message");
        }
    }
````
### Expected behavior

No warning

### Actual behavior

Warning S4457

### Known workarounds

Downgrade to 8.48 or suppress the warning

### Related information

* Project Type NET Standard 2.1 class library
* Visual Studio 16.4.1
* SonarScanner for .NET version 8.49.8.49.0.57237
* Windows Server 2022

### PR DESCRIPTION
**Pull request by gregory-paidis-sonarsource**
Fixes #6449

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInMethodWalker.cs [hunk 1]
```diff
@@ -61,16 +61,22 @@ public override void VisitThrowStatement(ThrowStatementSyntax node)
...
 
         public override void VisitInvocationExpression(InvocationExpressionSyntax node)
         {
             if (node.IsMemberAccessOnKnownType("ThrowIfNull", KnownType.System_ArgumentNullException, semanticModel))
             {
+                // "ThrowIfNull" returns void so it cannot be an argument. We can stop.
                 argumentExceptionLocations.Add(node.GetLocation());
             }
-            // "ThrowIfNull" returns void so it cannot be an argument. We can stop.
+            else
```
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInMethodWalker.cs:75]: Without this `else` block, the visit recursion stops for this node. 
This means that for something like:
```csharp
whatever.Invocation()
```
The `whatever.Invocation` syntax node is not going to keep visiting inside the `whatever` node, in which case if `whatever` contains an `AwaitExpression`, we will not spot it to stop the SyntaxWalker via:
```csharp
public override void VisitAwaitExpression(AwaitExpressionSyntax node) =>keepWalking = false;
```
Which produces this the FP in the testcases.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInMethodWalker.cs [hunk 2]
```diff
@@ -61,16 +61,22 @@ public override void VisitThrowStatement(ThrowStatementSyntax node)
             {
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInMethodWalker.cs:61]: This whole class is actually a helper and not a rule, so I suggest moving it to helpers.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInMethodWalker.cs (hunk 1 above)
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInMethodWalker.cs:75]: Nitpick and completely optional, as you will "work" on `ParameterValidationInYieldShouldBeWrapped` by adding test cases I would also clean up the implementation of the rule while you have the context.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
A few more modifications requested

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInMethodWalker.cs [hunk 3]
```diff
@@ -61,16 +61,22 @@ public override void VisitThrowStatement(ThrowStatementSyntax node)
...
             if (node.IsMemberAccessOnKnownType("ThrowIfNull", KnownType.System_ArgumentNullException, semanticModel))
             {
+                // "ThrowIfNull" returns void so it cannot be an argument. We can stop.
                 argumentExceptionLocations.Add(node.GetLocation());
             }
-            // "ThrowIfNull" returns void so it cannot be an argument. We can stop.
+            else
+            {
+                // Need to check the children of this node because of the pattern (await SomeTask()).Invocation()
+                DefaultVisit(node);
```
[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInMethodWalker.cs]: good catch, I missed that :(

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInAsyncShouldBeWrapped.cs [hunk 4]
```diff
@@ -41,13 +41,24 @@ public sealed class ParameterValidationInAsyncShouldBeWrapped : SonarDiagnosticA
...
                     if (walker.ArgumentExceptionLocations.Any())
                     {
                         c.ReportIssue(Rule.CreateDiagnostic(c.Compilation, method.Identifier.GetLocation(), walker.ArgumentExceptionLocations));
                     }
                 },
                 SyntaxKind.MethodDeclaration);
+
+        private sealed class ParameterValidationInAsyncWalker : ParameterValidationInMethodWalker
+        {
+            public ParameterValidationInAsyncWalker(SemanticModel semanticModel)
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInAsyncShouldBeWrapped.cs:55]: NITPICK:
```suggestion
            public ParameterValidationInAsyncWalker(SemanticModel semanticModel) : base(semanticModel) { }
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInYieldShouldBeWrapped.cs [hunk 5]
```diff
@@ -44,13 +42,12 @@ protected override void Initialize(SonarAnalysisContext context)
...
                     {
-                        c.ReportIssue(rule.CreateDiagnostic(c.Compilation, methodDeclaration.Identifier.GetLocation(), additionalLocations: walker.ArgumentExceptionLocations));
+                        c.ReportIssue(Rule.CreateDiagnostic(c.Compilation, methodDeclaration.Identifier.GetLocation(), additionalLocations: walker.ArgumentExceptionLocations));
                     }
                 },
                 SyntaxKind.MethodDeclaration);
-        }
 
-        private class ParameterValidationInYieldWalker : ParameterValidationInMethodWalker
+        private sealed class ParameterValidationInYieldWalker : ParameterValidationInMethodWalker
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInYieldShouldBeWrapped.cs:50]: NITPICK:
The constructor could be in a single line.

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**